encodings/fastlanes/src/bitpacking/compress.rs

```diff
@@ -22,28 +23,51 @@ pub fn bitpack_encode(array: PrimitiveArray, bit_width: usize) -> VortexResult<B
 
     if bit_width >= array.ptype().bit_width() {
         // Nothing we can do
-        vortex_bail!(
-            "Cannot pack -- specified bit width is greater than or equal to the type's bit width"
-        )
+        vortex_bail!("Cannot pack -- specified bit width is greater than or equal to raw bit width")
     }
 
     let packed = bitpack(&array, bit_width)?;
     let patches = (num_exceptions > 0)
         .then(|| bitpack_patches(&array, bit_width, num_exceptions))
         .flatten();
 
-    BitPackedArray::try_new(packed, array.validity(), patches, bit_width, array.len())
+    BitPackedArray::try_new(
+        packed,
+        array.ptype().to_unsigned(),
+        array.validity(),
+        patches,
+        bit_width,
+        array.len(),
+    )
 }
 
-pub fn bitpack(parray: &PrimitiveArray, bit_width: usize) -> VortexResult<Array> {
+/// Bitpack a [PrimitiveArray] to the given width.
+///
+/// On success, returns a [Buffer] containing the packed data.
+pub fn bitpack(parray: &PrimitiveArray, bit_width: usize) -> VortexResult<Buffer> {
     // We know the min is > 0, so it's safe to re-interpret signed integers as unsigned.
     let parray = parray.reinterpret_cast(parray.ptype().to_unsigned());
     let packed = match_each_unsigned_integer_ptype!(parray.ptype(), |$P| {
-        PrimitiveArray::from(bitpack_primitive(parray.maybe_null_slice::<$P>(), bit_width))
+        let packed_vec: Vec<$P> = bitpack_primitive(parray.maybe_null_slice::<$P>(), bit_width);
+        let (ptr, len, cap) = packed_vec.into_raw_parts();
+        let ptr_u8: *mut u8 = ptr.cast();
+        let len_u8 = len * std::mem::size_of::<$P>();
+        let cap_u8 = cap * std::mem::size_of::<$P>();
+
+        // SAFETY: we are simply reinterpreting the packed_vec as a Vec<u8>.
+        let packed_u8 = unsafe {
+            Vec::from_raw_parts(ptr_u8, len_u8, cap_u8)
+        };
+
+        Buffer::from(packed_u8)
```

Comment:
trivial, but consider moving Lines 52-62 might to be inside `bitpack_primitive` (and make that fn return `Buffer`). better to avoid doing complicated logic within a macro block since IDEs generally don't work well in those contexts